### ISSUE TITLE
Spreed app visible before 2FA completed on webUI

### ISSUE THREAD

**Issue by grrdvnk** [nextcloud/server#24027]
### Steps to reproduce
1. Login to nextcloud 
2. On 2FA screen manually point browser to <nextcloudurl>/apps/spreed without entering any TOTP key

### Expected behaviour
Redirect to <nextcloudurl>/login/selectchallenge?redirect_url=/settings/users

### Actual behaviour
App screen is shown (<nextcloudurl>/apps/spreed) with menubar visible as if succesfully logged in. Though no actual data is shown (conversations/contacts/etc) and interactions will fail (creating new conversations/search for contacts/etc).

### Server configuration

**Operating system:** Debian 10

**Web server:** Apache 2.4.38-3

**Database:** MariaDB 10.3.25

**PHP version:** 7.4.12 FPM

**Nextcloud version:** 19.0.4.2

**Comment by kesselb** [comment]
cc @nextcloud/security

**Comment by acsfer** [comment]
Tested on NC 19.0.4.2 + nginx.

When manually changing the UTL to `/apps/spreed` while on the 2FA screen,
i'm being redirected to `/login/selectchallenge?redirect_url=/apps/files/`

**Comment by Chartman123** [comment]
I can confirm it in 20.0.1 running the official Docker images

**Comment by rullzer** [comment]
It is because the spreed page is actually a public page. So you do not need to be logged in properly to view it.
However all the other loading indeed fails

**Comment by grrdvnk** [comment]
Sounds plausible. However it seems to be to inconsistent to be true: before the first authentication step (username/password) the spreed page is not public at all. Where would we benefit if it would suddenly become public in bewteen first & second step authentication? And I do see my personal avatar showing up... 

Also @acsfer is not affected and sees the expected behaviour (redirect to the 2FA login page). Maybe it's an apache2 combination bug?
I'll see if I can test it on a nginx instance. But first I'll run an update tonight (just noticed NC 19.0.5 has arrived, as well as Talk/spreed 9.0.6).

**Comment by rullzer** [comment]
https://github.com/nextcloud/spreed/blob/master/lib/Controller/PageController.php#L169

The reason when you are not logged in it works is because we then try to find your token. And if we fail it does redirects.

However when you just did not do 2fa yet. It will happily let you trough. Since your session has started. It is just lacking 2fa confirmation.
As you can see other requests you do are failing hard.

**Comment by rullzer** [comment]
@danxuliu @nickvergessen one for you I think then

**Comment by nickvergessen** [comment]
Well what is an app supposed to check again? I think we need a more dedicated annotation then?
Being half-logged in should not be a valid state for PublicPage, but only allow routes which can complete the 2FA?

**Comment by szaimen** [comment]
I suppose this is still valid?

**Comment by grrdvnk** [comment]
Yes, still valid for current version — spreed 11.2.2 — on NC 21.0.2

**Comment by szaimen** [comment]
Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!

If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+
